src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1052,3 +1259,93 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
 
 	return true;
 }
+
+
+/*
+ *  TopLevelUnionQuery gets a queryTree and returns true if top level query
+ * contains a UNION set operation. Note that the function allows upper level
+ * queries with a single range table element. In other words, we allow top
+ * level unions being wrapped into aggregations queries and/or simple projection
+ * queries that only selects some fields from the lower level queries.
+ *
+ * If there exists joins before the set operations, the function returns false.
+ * Similarly, if the query does not contain any union set operations, the
+ * function returns false.
+ */
+static bool
+TopLevelUnionQuery(Query *queryTree)
+{
+	List *rangeTableList = queryTree->rtable;
+	Node *setOperations = queryTree->setOperations;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+
+	/* don't allow joines on top of unions */
+	if (list_length(rangeTableList) > 1)
```

Comment:
Hmm, it could be that the name of the function is a bit confusing. This function is looking for a query tree such that the query may have as many upper level queries above the `UNION` query, unless none of them is a join.

Is the function comment not clear enough for that?